### PR TITLE
Steps to support the Dolly model

### PR DESCRIPTION
**Pull request by devkral**
# What
- The current glob for model files is very restricted, I relaxed it a little so it could find the dolly model.
- support for ByteStorage was added. As far as I can see it is uint8.
- new: PretrainedVocab is added

# Why

I want to use the dolly model with llama.cpp. They use some ByteStorage stuff of torch

# Remaining issues

<del>The vocab file is in a completely different format to sentencepiece (it uses a pretrained tokenizer):</del>

<del>Somehow it must get converted or another Vocab must be added to convert.py </del>

- dolly uses a gpt_neox format which is different from what llama.cpp understands. Needs conversion

### CODE REVIEW

File: convert.py [hunk 1]
```diff
@@ -205,6 +207,32 @@ def __repr__(self) -> str:
...
+        for item in self.added_tokens_list:
+            score = -1000.0
+            yield item["content"].encode("utf-8"), score
+
+    def all_tokens(self) -> Iterable[Tuple[bytes, float]]:
+        yield from self.vocab_tokens()
+        yield from self.added_tokens()
+
+    def __repr__(self) -> str:
+        return f"<SentencePieceVocab with {self.vocab_size_base} base tokens and {len(self.added_tokens_list)} added tokens>"
```
Comment (j-f1) [inline convert.py:233]:
```suggestion
        return f"<PretrainedVocab with {self.vocab_size_base} base tokens and {len(self.added_tokens_list)} added tokens>"
```

**Comment by ggerganov** [comment]
For a GPT-NeoX implementation using `ggml` see the [StableLM example](https://github.com/ggerganov/ggml/tree/master/examples/stablelm).
It will require some extra work to integrate this in `llama.cpp`.

But before doing that, we need to add a `ggml` example for Dolly and make sure that it works correctly.

**Comment by devkral** [comment]
nice, sry for ghost posting but why llama.cpp exists if you have the ggml repo?

I am a beginner in terms of AI.

**Comment by mikev-db** [comment]
> For a GPT-NeoX implementation using `ggml` see the [StableLM example](https://github.com/ggerganov/ggml/tree/master/examples/stablelm). It will require some extra work to integrate this in `llama.cpp`.
> 
> But before doing that, we need to add a `ggml` example for Dolly and make sure that it works correctly.

Hi @ggerganov I made an example here, I can PR it if it looks good. 

https://github.com/mverrilli/ggml/tree/dolly-v2/examples/dolly-v2

**Comment by mverrilli** [comment]
Hi @ggerganov 
 
> But before doing that, we need to add a `ggml` example for Dolly and make sure that it works correctly.

I created this one.  I can PR it if it is ok. 

https://github.com/mverrilli/ggml/tree/dolly-v2/examples/dolly-v2

**Comment by ggerganov** [comment]
@devkral

Yes, please open a PR.
In your experience, do the `ggml` results look OK when you compare to the reference Python implementation?
I'll probably do some more rigorous testing later, but would like to get an additional opinion.

**Comment by j-f1** [comment]
> nice, sry for ghost posting but why llama.cpp exists if you have the ggml repo?
> 
> I am a beginner in terms of AI.

GGML is a general purpose matrix API that doesn’t include support for running specific models directly (I believe). This repo exists to use GGML to implement the specific structures of LLaMA.

**Comment by devkral** [comment]
The ggml  dolly example results look good.

**Comment by xingchensong** [comment]
Hi teams， any update？

**Comment by devkral** [comment]
wrong repository

**Comment by xingchensong** [comment]
> wrong repository

is ggml the right one？

**Pull request closed without merging**